lib/extensions/postgres_cdc_rls/replication_poller.ex

```diff
@@ -99,79 +94,21 @@ defmodule Extensions.PostgresCdcRls.ReplicationPoller do
     cancel_timer(poll_ref)
     cancel_timer(retry_ref)
 
-    try do
-      {time, response} =
-        :timer.tc(Replications, :list_changes, [
-          conn,
-          slot_name,
-          publication,
-          max_changes,
-          max_record_bytes
-        ])
-
-      Realtime.Telemetry.execute(
-        [:realtime, :replication, :poller, :query, :stop],
-        %{duration: time},
-        %{tenant: tenant}
-      )
-
-      response
-    catch
-      {:error, reason} ->
-        {:error, reason}
-    end
-    |> case do
-      {:ok,
-       %Postgrex.Result{
-         columns: ["wal", "is_rls_enabled", "subscription_ids", "errors"] = columns,
-         rows: [_ | _] = rows,
-         num_rows: rows_count
-       }} ->
-        Enum.reduce(rows, [], fn row, acc ->
-          columns
-          |> Enum.zip(row)
-          |> generate_record()
-          |> case do
-            nil ->
-              acc
-
-            record_struct ->
-              [record_struct | acc]
-          end
-        end)
-        |> Enum.reverse()
-        |> Enum.each(fn change ->
-          Phoenix.PubSub.broadcast_from(
-            PubSub,
-            self(),
-            "realtime:postgres:" <> tenant,
-            change,
-            MessageDispatcher
-          )
-        end)
-
-        {:ok, rows_count}
-
-      {:ok, _} ->
-        {:ok, 0}
-
-      {:error, reason} ->
-        {:error, reason}
-    end
-    |> case do
-      {:ok, rows_num} ->
-        backoff = Backoff.reset(backoff)
-
-        poll_ref =
-          if rows_num > 0 do
-            send(self(), :poll)
-            nil
-          else
-            Process.send_after(self(), :poll, poll_interval_ms)
-          end
-
-        {:noreply, %{state | backoff: backoff, poll_ref: poll_ref}}
+    with args <- [conn, slot_name, publication, max_changes, max_record_bytes],
+         {time, list_changes} <- :timer.tc(Replications, :list_changes, args),
+         _ <- record_list_changes_telemetry(time, tenant),
+         {:ok, row_count} <- handle_list_changes_result(list_changes, tenant),
+         backoff <- Backoff.reset(backoff) do
+      pool_ref =
+        if row_count == 0 do
+          send(self(), :poll)
+          nil
+        else
+          Process.send_after(self(), :poll, poll_interval_ms)
+        end
```

Comment:
Here, if the rows are not empty, we should poll immediately to retrieve the rest of the changes. If there are no rows, we can wait

```suggestion
        if row_count > 0 do
          send(self(), :poll)
          nil
        else
          Process.send_after(self(), :poll, poll_interval_ms)
        end
```